contracts/SpokePool.sol

```diff
@@ -1,16 +1,176 @@
+        require(destinationToken != address(0), "Invalid spoke pool");
+        whitelistedDestinationRoutes[originToken][destinationChainId] = DestinationToken({
+            token: destinationToken,
+            isWethToken: isWethToken,
+            depositsEnabled: true
+        });
+
+        emit WhitelistRoute(originToken, destinationChainId, destinationToken, isWethToken);
+    }
+
+    /**
+     * @notice Enable/disable deposits for a whitelisted origin token.
+     */
+    function _setEnableDeposits(
+        address originToken, 
+        uint256 destinationChainId, 
+        bool depositsEnabled
+    ) internal onlyWhitelistedPath(originToken, destinationChainId) {
+        whitelistedDestinationRoutes[originToken][destinationChainId].depositsEnabled = depositsEnabled;
+        emit DepositsEnabled(originToken, destinationChainId, depositsEnabled);
+    }
+
+    /**************************************
+     *         DEPOSITOR FUNCTIONS        *
+     **************************************/
+
+    /**
+     * @notice Called by user to bridge funds from origin to destination chain.
+     * @dev The caller must first approve this contract to spend `amount` of `originToken`.
+     */
     function deposit(
         address originToken,
+        uint256 destinationChainId,
         uint256 amount,
         address recipient,
-        uint256 destinationChainId,
-        uint256 relayerFee
-    ) public {}
+        uint64 relayerFeePct,
+        uint64 quoteTimestamp
+    ) public payable onlyIfDepositsEnabled(originToken, destinationChainId) {
+        // We limit the relay fees to prevent the user spending all their funds on fees.
+        require(relayerFeePct <= 0.5e18, "invalid relayer fee");
+        // Note We assume that L2 timing cannot be compared accurately and consistently to L1 timing. Therefore, 
+        // `block.timestamp` is different from the L1 EVM's. Therefore, the quoteTimestamp must be within 10
+        // mins of the current time to allow for this variance.
+        // Note also that `quoteTimestamp` cannot be less than 10 minutes otherwise the following arithmetic can result
+        // in underflow. This isn't a problem as the deposit will revert, but the error might be unexpected for clients.
+        // Consider requiring `quoteTimestamp >= 10 minutes`.
+        require(
+            getCurrentTime() >= quoteTimestamp - 10 minutes && 
+            getCurrentTime() <= quoteTimestamp + 10 minutes &&
+            quoteTimestamp >= deploymentTime,
+            "invalid quote time"
+        );
+        DestinationToken memory tokenPath = whitelistedDestinationRoutes[originToken][destinationChainId];
+        // If the address of the destination token is a WETH contract and there is a msg.value with the transaction 
+        // then the user is sending ETH. In this case, the ETH should be deposited to WETH.
+        if (tokenPath.isWethToken && msg.value > 0) {
```

Comment:
As mentioned above, I think it's a bit odd that we're using the token path to determine whether the _deposited token_ is WETH, since that seems to be something that only depends on the origin chain.